File: clients/ios/Lextures/Features/Navigation/GlobalDrawer.swift
import SwiftUI

/// App-wide navigation drawer mirroring the Lextures web sidebar: brand header,
/// search, and grouped destinations. Selecting a row switches the top-level pane.
struct GlobalDrawer: View {
    @Environment(AppShellModel.self) private var shell
    @Environment(\.colorScheme) private var colorScheme

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            header
            searchButton
                .padding(.horizontal, 14)
                .padding(.bottom, 8)

            ScrollView {
                VStack(alignment: .leading, spacing: 4) {
                    // Pinned course covers sit directly under search, above the nav (web-parity).
                    pinnedTiles
                    ForEach(Array(shell.globalDrawerGroups.enumerated()), id: \.offset) { _, group in
                        if let title = group.title {
                            DrawerGroupHeader(title: title)
                        }
                        ForEach(group.items) { item in
                            DrawerRow(
                                label: item.label,
                                systemImage: item.systemImage,
                                selected: shell.rootDestination == item,
                                badge: item.showsInboxBadge ? shell.unreadInbox : 0
                            ) {
                                shell.select(item)
                            }
                        }
                    }
                }
                .padding(.horizontal, 10)
                .padding(.bottom, 24)
            }
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topLeading)
        .background(LexturesTheme.sceneBackground(for: colorScheme).ignoresSafeArea())
    }

    @ViewBuilder
    private var pinnedTiles: some View {
        if !shell.pinnedCourses.isEmpty {
            LazyVGrid(
                columns: [GridItem(.adaptive(minimum: 76, maximum: 120), spacing: 8)],
                alignment: .leading,
                spacing: 8
            ) {
                ForEach(shell.pinnedCourses) { course in
                    let active = shell.activeCourse?.courseCode == course.courseCode
                    Button {
                        shell.closeDrawer()
                        shell.openDeepLink(.course(code: course.courseCode, section: nil, itemId: nil))
                    } label: {
                        CourseHeroImage(
                            urlString: course.heroImageUrl,
                            fallbackKey: course.courseCode,
                            height: nil
                        )
                        .frame(height: 52)
                        .frame(maxWidth: .infinity)
                        .clipShape(RoundedRectangle(cornerRadius: 14, style: .continuous))
                        .overlay(
                            RoundedRectangle(cornerRadius: 14, style: .continuous)
                                .stroke(
                                    active ? LexturesTheme.accent(for: colorScheme) : LexturesTheme.fieldBorder(for: colorScheme),
                                    lineWidth: active ? 2 : 1
                                )
                        )
                    }
                    .buttonStyle(.plain)
                    .accessibilityLabel(course.displayTitle)
                }
            }
            .padding(.horizontal, 2)
            .padding(.top, 2)
            .padding(.bottom, 8)
        }
    }

    private var header: some View {
        HStack(spacing: 10) {
            BrandLogoView(maxHeight: 30)
                .frame(width: 34, height: 34)
            Text("Lextures")
                .font(LexturesTheme.displayFont(20))
                .foregroundStyle(LexturesTheme.textPrimary(for: colorScheme))
            Spacer(minLength: 0)
            LMSAvatarButton(size: 34)
        }
        .padding(.horizontal, 16)
        .padding(.top, 16)
        .padding(.bottom, 14)
    }

    private var searchButton: some View {
        Button {
            shell.closeDrawer()
            shell.showUniversalSearch = true
        } label: {
            HStack(spacing: 10) {
                Image(systemName: "magnifyingglass")
                    .font(.subheadline)
                    .foregroundStyle(LexturesTheme.textSecondary(for: colorScheme))
                Text(L.text("mobile.ia.search"))
                    .font(.subheadline)
                    .foregroundStyle(LexturesTheme.textSecondary(for: colorScheme))
                Spacer(minLength: 0)
            }
            .padding(.horizontal, 12)
            .padding(.vertical, 10)
            .background(LexturesTheme.cardBackground(for: colorScheme))
            .clipShape(RoundedRectangle(cornerRadius: 12, style: .continuous))
            .overlay(
                RoundedRectangle(cornerRadius: 12, style: .continuous)
                    .stroke(LexturesTheme.fieldBorder(for: colorScheme), lineWidth: 1)
            )
        }
        .buttonStyle(.plain)
    }
}

// MARK: - Shared drawer chrome

/// Small-caps grouping header used by both drawers.
struct DrawerGroupHeader: View {
    @Environment(\.colorScheme) private var colorScheme
    let title: String

    var body: some View {
        Text(title.uppercased())
            .font(.caption2.weight(.semibold))
            .tracking(0.6)
            .foregroundStyle(LexturesTheme.textSecondary(for: colorScheme))
            .padding(.horizontal, 12)
            .padding(.top, 14)
            .padding(.bottom, 4)
            .frame(maxWidth: .infinity, alignment: .leading)
    }
}

/// A single tappable drawer row with optional selection pill and unread badge.
struct DrawerRow: View {
    @Environment(\.colorScheme) private var colorScheme
    let label: String
    let systemImage: String
    var selected: Bool = false
    var badge: Int = 0
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            HStack(spacing: 12) {
                Image(systemName: systemImage)
                    .font(.system(size: 16, weight: .medium))
                    .frame(width: 24, alignment: .center)
                    .foregroundStyle(
                        selected
                            ? LexturesTheme.accent(for: colorScheme)
                            : LexturesTheme.textSecondary(for: colorScheme)
                    )
                Text(label)
                    .font(.subheadline.weight(selected ? .semibold : .regular))
                    .foregroundStyle(LexturesTheme.textPrimary(for: colorScheme))
                Spacer(minLength: 0)
                if badge > 0 {
                    Text(badge > 99 ? "99+" : "\(badge)")
                        .font(.caption2.weight(.bold))
                        .foregroundStyle(.white)
                        .padding(.horizontal, 6)
                        .padding(.vertical, 2)
                        .background(LexturesTheme.coral)
                        .clipShape(Capsule())
                }
            }
            .padding(.horizontal, 12)
            .padding(.vertical, 11)
            .background(
                RoundedRectangle(cornerRadius: 12, style: .continuous)
                    .fill(selected
                        ? LexturesTheme.brandTeal.opacity(colorScheme == .dark ? 0.22 : 0.16)
                        : .clear)
            )
        }
        .buttonStyle(.plain)
        .accessibilityAddTraits(selected ? [.isSelected] : [])
    }
}

File: clients/android/app/src/main/kotlin/com/lextures/android/features/navigation/GlobalDrawer.kt
package com.lextures.android.features.navigation

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.foundation.layout.height
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Search
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.lextures.android.R
import com.lextures.android.core.design.LexturesColors
import com.lextures.android.core.design.cardBackground
import com.lextures.android.core.design.fieldBorder
import com.lextures.android.core.design.isDarkTheme
import com.lextures.android.core.design.sceneBackground
import com.lextures.android.core.design.textPrimary
import com.lextures.android.core.design.textSecondary
import com.lextures.android.core.i18n.L
import com.lextures.android.core.lms.CourseSummary
import com.lextures.android.core.lms.resolvedInitials
import com.lextures.android.core.design.LocalUIModeStore
import com.lextures.android.core.design.UIMode
import com.lextures.android.core.design.UIModeLogic
import com.lextures.android.core.navigation.MobileDestinations
import com.lextures.android.features.home.CourseHeroImage
import com.lextures.android.features.home.HomeShellState
import com.lextures.android.features.home.LmsAvatarChip

/**
 * App-wide navigation drawer mirroring the Lextures web sidebar: brand header, search,
 * and grouped destinations. Selecting a row switches the top-level pane.
 */
@Composable
fun GlobalDrawer(shell: HomeShellState, accessToken: String?, modifier: Modifier = Modifier) {
    val uiModeStore = LocalUIModeStore.current
    val uiMode = uiModeStore.effectiveMode(shell.activeRoleContext)
    Column(
        modifier = modifier
            .fillMaxSize()
            .background(sceneBackground()),
    ) {
        // Header: brand + avatar.
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(start = 16.dp, end = 16.dp, top = 16.dp, bottom = 14.dp),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(10.dp),
        ) {
            Icon(
                painter = painterResource(R.drawable.launch_logo),
                contentDescription = null,
                tint = androidx.compose.ui.graphics.Color.Unspecified,
                modifier = Modifier.size(32.dp),
            )
            Text("Lextures", fontSize = 20.sp, fontWeight = FontWeight.Bold, color = textPrimary())
            androidx.compose.foundation.layout.Spacer(Modifier.weight(1f))
            LmsAvatarChip(
                initials = shell.accountProfile?.resolvedInitials() ?: shell.profile?.initials ?: "··",
                onClick = { shell.select(com.lextures.android.core.navigation.RootDestination.Profile) },
                size = 34,
                avatarUrl = shell.accountProfile?.avatarUrl,
            )
        }

        if (MobileDestinations.showsUniversalSearch(uiMode)) {
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(horizontal = 14.dp)
                    .clip(RoundedCornerShape(12.dp))
                    .background(cardBackground())
                    .border(1.dp, fieldBorder(), RoundedCornerShape(12.dp))
                    .clickable {
                        shell.closeDrawer()
                        shell.showUniversalSearch = true
                    }
                    .padding(horizontal = 12.dp, vertical = 10.dp),
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(10.dp),
            ) {
                Icon(Icons.Default.Search, contentDescription = null, tint = textSecondary(), modifier = Modifier.size(18.dp))
                Text(L.text(R.string.mobile_ia_search), color = textSecondary(), fontSize = 14.sp)
            }
        }
        // (pinned courses rendered inside the grouped list below)

        Column(
            modifier = Modifier
                .fillMaxWidth()
                .verticalScroll(rememberScrollState())
                .padding(horizontal = 10.dp, vertical = 8.dp),
        ) {
            // Pinned course covers sit directly under search, above the nav (web-parity).
            if (shell.pinnedCourses.isNotEmpty()) {
                PinnedCourseTiles(
                    courses = shell.pinnedCourses,
                    activeCourseCode = shell.activeCourse?.courseCode,
                    accessToken = accessToken,
                    onOpen = { course ->
                        shell.closeDrawer()
                        shell.openDeepLink(
                            com.lextures.android.core.routing.DeepLinkDestination.Course(course.courseCode),
                        )
                    },
                )
            }
            shell.globalDrawerGroups.forEach { group ->
                group.titleRes?.let { DrawerGroupHeader(drawerString(it)) }
                group.items.forEach { item ->
                    DrawerRow(
                        label = UIModeLogic.drawerLabel(item, uiMode),
                        icon = rootDestinationIcon(item),
                        selected = shell.rootDestination == item,
                        onClick = { shell.select(item) },
                        badge = if (item.showsInboxBadge) shell.unreadInbox else 0,
                        uiMode = uiMode,
                    )
                }
            }
        }
    }
}

/** Pinned course covers as a wrapping row of rounded tiles — no labels (web-parity). */
@OptIn(androidx.compose.foundation.layout.ExperimentalLayoutApi::class)
@Composable
private fun PinnedCourseTiles(
    courses: List<CourseSummary>,
    activeCourseCode: String?,
    accessToken: String?,
    onOpen: (CourseSummary) -> Unit,
) {
    androidx.compose.foundation.layout.FlowRow(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 2.dp, vertical = 2.dp),
        horizontalArrangement = Arrangement.spacedBy(8.dp),
        verticalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        courses.forEach { course ->
            val active = course.courseCode == activeCourseCode
            Box(
                modifier = Modifier
                    .width(88.dp)
                    .height(52.dp)
                    .clip(RoundedCornerShape(14.dp))
                    .border(
                        width = if (active) 2.dp else 1.dp,
                        color = if (active) LexturesColors.BrandTeal else fieldBorder(),
                        shape = RoundedCornerShape(14.dp),
                    )
                    .clickable { onOpen(course) }
                    .semantics { contentDescription = course.displayTitle },
            ) {
                CourseHeroImage(
                    url = course.heroImageUrl,
                    fallbackKey = course.courseCode,
                    accessToken = accessToken,
                    height = null,
                )
            }
        }
    }
}
